config/axi_memory_map/create_linker_script.py

```diff
@@ -67,90 +66,105 @@
+
+DEVICE_NAME = 0
+DEVICE_ORIGIN = 1
+DEVICE_LENGTH = 2
+
+memory_block_list = []
+peripheral_list = []
 counter = 0
-PERIPHERALS_LENGTH = 0
-BRAM_LENGTH = 0
-DDR_LENGTH = 0
-HBM_LENGTH = 0
-BRAM_END = 0
-DDR_END = 0
-HBM_END = 0
-for name in RANGE_NAMES:
-	match name:
-		# BRAM
-		case "BRAM":
-			BRAM_ORIGIN = int(RANGE_BASE_ADDR[counter], 16)
-			BRAM_LENGTH = 2 << RANGE_ADDR_WIDTH[counter]
-			BRAM_END 	= BRAM_ORIGIN + BRAM_LENGTH
-		# DDR
-		case "DDR":
-			DDR_ORIGIN = int(RANGE_BASE_ADDR[counter], 16)
-			DDR_LENGTH = 2 << RANGE_ADDR_WIDTH[counter]
-			DDR_END    = DDR_ORIGIN + DDR_LENGTH
-		# HBM
-		case "HBM":
-			HBM_ORIGIN = int(RANGE_BASE_ADDR[counter], 16)
-			HBM_LENGTH = 2 << RANGE_ADDR_WIDTH[counter]
-			HBM_END    = HBM_ORIGIN + HBM_LENGTH
+for device in RANGE_NAMES:
+	match device:
+		# memory blocks
+		case "BRAM" | "DDR" | "HBM":
+			memory_block_list.append([device, int(RANGE_BASE_ADDR[counter], 16), 2 << RANGE_ADDR_WIDTH[counter]])
+
 		# Peripherals
 		case _:
-			# Max reduce
-			PERIPHERALS_LENGTH = max(PERIPHERALS_LENGTH, 2 << RANGE_ADDR_WIDTH[counter])
+			peripheral_list.append([device, int(RANGE_BASE_ADDR[counter], 16), 2 << RANGE_ADDR_WIDTH[counter]])
 
 	# Increment counter
 	counter += 1
 
-# Peripherals base, soon as the memory space ends
-PERIPHERALS_ORIGIN = max(BRAM_END, DDR_END, HBM_END)
-
-# Peripherals length,
-PERIPHERALS_LENGTH = 0
-for i in range(1, len(RANGE_ADDR_WIDTH)):
-	PERIPHERALS_LENGTH += (2 << RANGE_ADDR_WIDTH[i])
+###############################
+# Generate Linker Script File #
+###############################
 
-#################
-# Write to file #
-#################
-# Open output file
+# Create the Linker Script File
 fd = open(ld_file_name,  "w")
 
 # Write header
 fd.write("/* This file is auto-generated with " + os.path.basename(__file__) + " */\n")
 
-# BLOCKS
+# Generate the memory blocks layout
 fd.write("\n")
 fd.write("/* Memory blocks */\n")
 fd.write("MEMORY\n")
 fd.write("{\n")
-# Memories
-if BRAM_LENGTH != 0:
-	fd.write("\tBRAM (xrw) : ORIGIN = " + hex(BRAM_ORIGIN) + ",  LENGTH = " + hex(BRAM_LENGTH) + "\n")
-if DDR_LENGTH != 0:
-	fd.write("\tDDR (xrw) : ORIGIN = " + hex(DDR_ORIGIN) + ",  LENGTH = " + hex(DDR_LENGTH) + "\n")
-if HBM_LENGTH != 0:
-	fd.write("\tHBM (xrw) : ORIGIN = " + hex(HBM_ORIGIN) + ",  LENGTH = " + hex(HBM_LENGTH) + "\n")
-# Peripherals
-fd.write("\tPERIPHERALS (rw) : ORIGIN = " + hex(PERIPHERALS_ORIGIN) + ",  LENGTH = " + hex(PERIPHERALS_LENGTH) + "\n")
+
+for block in memory_block_list:
+	fd.write("\t" + block[DEVICE_NAME] + " (xrw) : ORIGIN = 0x" + format(block[DEVICE_ORIGIN], "08x") + ",  LENGTH = " + hex(block[DEVICE_LENGTH]) + "\n")
 fd.write("}\n")
 
-# SECTIONS
+# Generate symbols from peripherals
+fd.write("\n")
+fd.write("/* Peripherals symbols */\n")
+for peripheral in peripheral_list:
+	fd.write("_peripheral_" + peripheral[DEVICE_NAME] + "_start = 0x" + format(peripheral[DEVICE_ORIGIN], "08x") + ";\n")
```

Comment:
format string `08x` assumes 32-bit address space. Should we not prepare for RV64 here? I.e. simply by using `016x`

Reply:
done